content/courses/advanced-ada/chapters/types.rst

```diff
@@ -6,14 +6,432 @@ Types
++                       +-------------------------+---------------------------+
+|                       | :ada:`Wide_Wide_Image`  | :ada:`Wide_Wide_String`   |
++-----------------------+-------------------------+---------------------------+
+| Conversion to subtype | :ada:`Value`            | :ada:`String`             |
++                       +-------------------------+---------------------------+
+|                       | :ada:`Wide_Value`       | :ada:`Wide_String`        |
++                       +-------------------------+---------------------------+
+|                       | :ada:`Wide_Wide_Value`  | :ada:`Wide_Wide_String`   |
++-----------------------+-------------------------+---------------------------+
+
+We discuss more about :ada:`Wide_String` and :ada:`Wide_Wide_String` in
+:ref:`another section <Wide_Wide_Strings>`.
+
+Width attribute
+~~~~~~~~~~~~~~~
+
+When converting a value to a string by using the :ada:`Image` attribute, we get
+a string with variable width. We can assess the maximum width of that string
+for a specific subtype by using the :ada:`Width` attribute. For example,
+:ada:`Integer'Width` gives us the maximum width returned by the :ada:`Image`
+attribute when converting a value of :ada:`Integer` type to a string of
+:ada:`String` type.
+
+This attribute is useful when we're using bounded strings in our code to store
+the string returned by the :ada:`Image` attribute. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Width_Attr
+
+    with Ada.Text_IO;         use Ada.Text_IO;
+    with Ada.Strings;         use Ada.Strings;
+    with Ada.Strings.Bounded;
+
+    procedure Show_Width_Attr is
+       package B_Str is new
+         Ada.Strings.Bounded.Generic_Bounded_Length (Max => Integer'Width);
+       use B_Str;
+
+       Str_I : Bounded_String;
+
+       I : constant Integer := 42;
+       J : constant Integer := 103;
+    begin
+       Str_I := To_Bounded_String (I'Image);
+       Put_Line ("Value:         " & To_String (Str_I));
+       Put_Line ("String Length: " & Length (Str_I)'Image);
+       Put_Line ("----");
+
+       Str_I := To_Bounded_String (J'Image);
+       Put_Line ("Value:         " & To_String (Str_I));
+       Put_Line ("String Length: " & Length (Str_I)'Image);
+    end Show_Width_Attr;
+
+In this example, we're storing the string returned by :ada:`Image` in the
+:ada:`Str_I` variable of :ada:`Bounded_String` type.
+
+Similar to the :ada:`Image` and :ada:`Value` attributes, the :ada:`Width`
+attribute is also available for string types other than :ada:`String`. In fact,
+we can use:
+
+- the :ada:`Wide_Width` attribute for strings returned by :ada:`Wide_Image`;
+  and
+
+- the :ada:`Wide_Wide_Width` attribute for strings returned by
+  :ada:`Wide_Wide_Image`.
+
+Base
+~~~~
+
+The :ada:`Base` attribute gives us the unconstrained version of a subtype. As
+an example, let's say we declared a subtype of the :ada:`Integer` type named
+:ada:`One_To_Ten`:
```

Comment:
This is a subtle topic! 
Background: 
"Integer" is actually a subtype too. (These used to be known as "first named subtypes" -- the terminology has changed -- but the term does suggest correctly that they are a subtype of something that does not have an Ada name, and that is not constrained. That anonymous "type" is the hardware's representation for the numeric type as chosen by the compiler based on the requested range (for the signed integer types) or digits of precision (for floating-point types). Let's focus on signed integers. The other numerics work the same way. When we declare a signed integer type we have to specify the required range, statically. If the compiler cannot find a hardware-defined/supported signed integer type with at least the range requested, the compilation is rejected. Otherwise, the compiler maps the named Ada type to the hardware "type", presumably choosing the smallest one that supports the requested range. (That's why the range has to be static in the source code, unlike for explicit subtypes.) For any integer type named Foo, we can refer to that anonymous hardware type via Foo'Base.
(I've never used the term "first named subtype" in a training course and don't recommend it here either.) In effect, 

   type Integer is range L .. U;

is really as if we said this:

   subtype Integer is Some_Hardware_Type_With_Sufficient_Range range L .. U;

But that's impossible, so we just say that Integer is a type rather than a subtype.

So, what I suggest is 
"The :ada:`Base` attribute gives us the unconstrained version of a subtype." =>
"The :ada:`Base` attribute gives us the unconstrained underlying hardware representation selected for a given numeric type."

Reply:
I've integrated this change you suggested. However, your explanation about the hardware representation is interesting, so I'm asking myself whether we should add it to the section (maybe as an admonition). What do you think?

Reply:
I agree that it would be good to have the material included.

Reply:
Awesome! I added your explanation to the section and did some modifications here and there.